components/esp_eth/src/esp_eth_phy_ksz80xx.c

```diff
@@ -123,30 +58,23 @@ static esp_err_t ksz80xx_update_link_duplex_speed(phy_ksz80xx_t *ksz80xx)
     if (ksz80xx->link_status != link) {
         /* when link up, read negotiation result */
         if (link == ETH_LINK_UP) {
-            int op_mode = 0;
-            if (ksz80xx->vendor_model == KSZ8041_MODEL_ID) {
-                ksz8041_pc2r_reg_t pc2r;
-                ESP_GOTO_ON_ERROR(eth->phy_reg_read(eth, ksz80xx->addr, KSZ8041_PC2R_REG_ADDR, &(pc2r.val)), err, TAG, "read PC2R failed");
-                op_mode = pc2r.op_mode;
-            } else if (ksz80xx->vendor_model == KSZ8081_MODEL_ID) {
-                ksz8081_pc1r_reg_t pc1r;
-                ESP_GOTO_ON_ERROR(eth->phy_reg_read(eth, ksz80xx->addr, KSZ8081_PC1R_REG_ADDR, &(pc1r.val)), err, TAG, "read PC1R failed");
-                op_mode = pc1r.op_mode;
-            }
+            uint32_t reg_value = 0;
+            ESP_GOTO_ON_ERROR(eth->phy_reg_read(eth, ksz80xx->addr, ksz80xx->op_mode_reg, &reg_value), err, TAG, "read %#04x failed", ksz80xx->op_mode_reg);
```

Comment:
The only left concerns in the new implementation is about endianess of the read out register. As the new implementation uses bit shift operator instead of union-struct-pattern, it has to be validated if this line works as intended:
```uint8_t op_mode = (reg_value >> ksz80xx->op_mode_offset) & 0b111;```

Maybe i have time to test this on the real hardware in the next days.

Reply:
Testing it on real hardware would be highly appreciated!

Reply:
Just tested the alignment of the result returned by phy_reg_read.

The LSB of the register is aligned to the LSB of the uint32_t variable on the host architecture (esp32).
The upper 16 bit of the read value are unused.

So the implementation with the bitshift should work. Supposed, that the current implementation of the bitfield usage is correct.

Test-Code:
```
union {
    struct {
        uint32_t dis_data_scr : 1;   /* Disable Scrambler */
        uint32_t en_sqe_test : 1;    /* Enable SQE Test */
        uint32_t op_mode : 3;        /* Operation Mode */
        uint32_t phy_iso : 1;        /* PHY Isolate */
        uint32_t en_flow_ctrl : 1;   /* Enable Flow Control */
        uint32_t auto_nego_comp : 1; /* Auto Negotiation Complete */
        uint32_t en_jabber : 1;      /* Enable Jabber Counter */
        uint32_t irq_level : 1;      /* Interrupt Pin Active Level */
        uint32_t power_saving : 1;   /* Enable Powering Saving */
        uint32_t force_link : 1;     /* Force Link Pass */
        uint32_t energy_det : 1;     /* Presence of Signal on RX+/- Wire Pair */
        uint32_t pair_swap_dis : 1;  /* Disable Auto MDI/MDI-X */
        uint32_t mdix_select : 1;    /* MDI/MDI-X Select */
        uint32_t hp_mdix : 1;        /* HP Auto MDI/MDI-X Mode */
        uint32_t abc : 1;            /* HP Auto MDI/MDI-X Mode */
    };
    uint32_t val;
} t;
ESP_LOGE("test", "sizeof(t): %u", sizeof(t));

t.val = 1;
uint8_t bit = 1;
do {
    ESP_LOGE("test", "bit(%u) val(%#010X) | dis_data_scr(%u) en_sqe_test(%u) op_mode(%u) phy_iso(%u) en_flow_ctrl(%u) auto_nego_comp(%u) en_jabber(%u) irq_level(%u) power_saving(%u) force_link(%u) energy_det(%u) pair_swap_dis(%u) mdix_select(%u) hp_mdix(%u) abc(%u)",
                bit,  t.val,    t.dis_data_scr,  t.en_sqe_test,  t.op_mode,  t.phy_iso,  t.en_flow_ctrl,  t.auto_nego_comp,  t.en_jabber,  t.irq_level,  t.power_saving,  t.force_link,  t.energy_det,  t.pair_swap_dis,  t.mdix_select,  t.hp_mdix,  t.abc);
    t.val <<= 1;
    bit++;
} while (t.val);
```

Output (run on esp32):
```
[...] sizeof(t): 4
[...] bit(1)  val(0X00000001) | dis_data_scr(1) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)
[...] bit(2)  val(0X00000002) | dis_data_scr(0) en_sqe_test(1) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)
[...] bit(3)  val(0X00000004) | dis_data_scr(0) en_sqe_test(0) op_mode(1) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)
[...] bit(4)  val(0X00000008) | dis_data_scr(0) en_sqe_test(0) op_mode(2) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)
[...] bit(5)  val(0X00000010) | dis_data_scr(0) en_sqe_test(0) op_mode(4) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)
[...] bit(6)  val(0X00000020) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(1) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)
[...] bit(7)  val(0X00000040) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(1) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)
[...] bit(8)  val(0X00000080) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(1) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)
[...] bit(9)  val(0X00000100) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(1) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)
[...] bit(10) val(0X00000200) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(1) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)
[...] bit(11) val(0X00000400) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(1) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)  
[...] bit(12) val(0X00000800) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(1) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)  
[...] bit(13) val(0X00001000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(1) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)  
[...] bit(14) val(0X00002000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(1) mdix_select(0) hp_mdix(0) abc(0)  
[...] bit(15) val(0X00004000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(1) hp_mdix(0) abc(0)  
[...] bit(16) val(0X00008000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(1) abc(0)  
[...] bit(17) val(0X00010000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(1)
[...] bit(18) val(0X00020000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)  
[...] bit(19) val(0X00040000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)  
[...] bit(20) val(0X00080000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)  
[...] bit(21) val(0X00100000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)  
[...] bit(22) val(0X00200000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)  
[...] bit(23) val(0X00400000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)  
[...] bit(24) val(0X00800000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)  
[...] bit(25) val(0X01000000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)  
[...] bit(26) val(0X02000000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)  
[...] bit(27) val(0X04000000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)  
[...] bit(28) val(0X08000000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)  
[...] bit(29) val(0X10000000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)
[...] bit(30) val(0X20000000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)  
[...] bit(31) val(0X40000000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)  
[...] bit(32) val(0X80000000) | dis_data_scr(0) en_sqe_test(0) op_mode(0) phy_iso(0) en_flow_ctrl(0) auto_nego_comp(0) en_jabber(0) irq_level(0) power_saving(0) force_link(0) energy_det(0) pair_swap_dis(0) mdix_select(0) hp_mdix(0) abc(0)
```